File: src/example_cpp_node/src/subscriber_node.cpp
/**
 * @file subscriber_node.cpp
 * @brief Example ROS2 subscriber node using custom messages in C++
 */

#include <memory>

#include "example_msgs/msg/example_message.hpp"
#include "rclcpp/rclcpp.hpp"

class ExampleSubscriber : public rclcpp::Node {
 public:
  ExampleSubscriber() : Node("example_subscriber_cpp") {
    subscription_ = this->create_subscription<example_msgs::msg::ExampleMessage>(
        "example_topic_cpp", 10,
        std::bind(&ExampleSubscriber::topic_callback, this, std::placeholders::_1));
    RCLCPP_INFO(this->get_logger(), "Example Subscriber C++ Node started");
  }

 private:
  void topic_callback(const example_msgs::msg::ExampleMessage::SharedPtr msg) const {
    RCLCPP_INFO(this->get_logger(),
                "Received: name='%s', value=%d, timestamp=%f, is_active=%s",
                msg->name.c_str(), msg->value, msg->timestamp,
                msg->is_active ? "true" : "false");
  }

  rclcpp::Subscription<example_msgs::msg::ExampleMessage>::SharedPtr subscription_;
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ExampleSubscriber>());
  rclcpp::shutdown();
  return 0;
}

File: src/example_py_node/example_py_node/subscriber_node.py
#!/usr/bin/env python3
"""Example ROS2 subscriber node using custom messages."""

import rclpy
from example_msgs.msg import ExampleMessage
from rclpy.node import Node


class ExampleSubscriber(Node):
    """Subscribes to example messages and logs them."""

    def __init__(self):
        super().__init__("example_subscriber")
        self.subscription = self.create_subscription(
            ExampleMessage, "example_topic", self.listener_callback, 10
        )
        self.get_logger().info("Example Subscriber Node started")

    def listener_callback(self, msg):
        """Handle received messages."""
        self.get_logger().info(
            f'Received: name="{msg.name}", value={msg.value}, '
            f"timestamp={msg.timestamp}, is_active={msg.is_active}"
        )


def main(args=None):
    """Main entry point for the subscriber node."""
    rclpy.init(args=args)
    node = ExampleSubscriber()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
